A69-crl-enhancements.md

```diff
@@ -0,0 +1,222 @@
+
+
+## Implementation 
+### Go
+CRL checking in Go is entirely dependent upon setting the `tls.config.VerifyPeerCertificate` callback function to check CRLs. Currently, the Go implementation overrides this callback and from that calls `CheckChainRevocation(...)` which takes the certificate chain to be checked and the RevocationConfig. It gets the CRL of interest by calling `FetchIssuerCrl()` - we can use our provider in that function. Then it checks the cert chain against the relevant CRLs.
+
+We will add CRL Provider code to the advancedtls module. All CRL related code is currently in the advancedtls module, so this would keep it all in the same place.  Revocation configurations will be added to the existing configuration options in `RevocationConfig`. There will be checks to ensure a user can exclusively use the new provider options or the already-implemented X509 hash directory options. Configuring both is a mistake.
+```go
+// RevocationConfig contains options for CRL lookup.
+type RevocationConfig struct {
+	// RootDir is the directory to search for CRL files.
+	// Directory format must match OpenSSL X509_LOOKUP_hash_dir(3).
+	RootDir string
+	// AllowUndetermined controls if certificate chains with RevocationUndetermined
+	// revocation status are allowed to complete.
+	AllowUndetermined bool
+	// Cache will store CRL files if not nil, otherwise files are reloaded for every lookup.
+	Cache Cache
+
+  // crlProvider and RootDir/Cache cannot both be set
++ CrlProvider crlProvider
+}
+```
+
+The core CrlProvider interface will look as follows. Since x509 is a built in go library, we can safely use it in our API.
+```go
+type CRLProvider interface {
+	// Callers are expected to use the returned value as read-only.
+	CRL(cert x509.Certificate) (*CertificateListExt, error)
+
+	// Close cleans up resources allocated by the Provider.
+	Close()
+}
+```
+
+Regarding sharing between connections - since the providers are in a high-level configuration, the same provider should be used across many individual connections and handshakes. [Here is an example of how the credential reloading provider is configured](https://github.com/grpc/grpc-go/blob/master/security/advancedtls/examples/credential_reloading_from_files/server/main.go#L63). In this example, the server would use this provider for all connections.
+
+#### Directory Reloader Specifics
+![Directory Reloader Flow](A69_graphics/golang_reloader.png)
+```go
+type CrlDirectoryReloadingProvider struct {
+	ReloadInterval time.Duration
+	CrlDirectory string
```

Comment:
The resulting struct we implemented looks a little different (for example, most of the params are wrapped in Options struct), but I think this sample represents the overall idea quite good so I don't think we need to change it. @gtcooke94 WDYT?

Reply:
I think it's okay, this is meant to be a design idea, not precise and exact code